extcap_ot.py

```diff
@@ -0,0 +1,212 @@
+    if match and int(match.group(1)) >= 3:
+        args.append((Config.TAP.value, '--tap', 'IEEE 802.15.4 TAP (only for Wireshark3.0 and later)', 'IEEE 802.15.4 TAP', 'boolflag', '{default=yes}'))
+
+    for arg in args:
+        print('arg {number=%d}{call=%s}{display=%s}{tooltip=%s}{type=%s}%s' % arg)
+
+    values = values + [(Config.CHANNEL.value, '%d' % i, '%d' % i, 'true' if i == 11 else 'false') for i in range(11, 27)]
+
+    for value in values:
+        print('value {arg=%d}{value=%s}{display=%s}{default=%s}' % value)
+
+def extcap_dlts(interface):
+    """List DLTs for the given interface"""
+    print('dlt {number=195}{name=IEEE802_15_4_WITHFCS}{display=IEEE 802.15.4 with FCS}')
+    print('dlt {number=283}{name=IEEE802_15_4_TAP}{display=IEEE 802.15.4 TAP}')
+
+def serialopen(interface, console, log_file):
+    """
+    Open serial to indentify OpenThread sniffer
+    :param interface: string, eg: '/dev/ttyUSB0 - Zolertia Firefly platform', '/dev/ttyACM1 - nRF52840 OpenThread Device'
+    """
+    sys.stdout = log_file
+    sys.stderr = log_file
+    interface = str(interface).split()[0]
+    baudrate = None
+
+    for speed in COMMON_BAUDRATE:
+        stream = StreamOpen('u', interface, False, baudrate=speed)
+
+        wpan_api = WpanApi(stream, nodeid=DEFAULT_NODEID)
+        # result is None for NCP, not None for RCP
+        wpan_api.prop_set_value(SPINEL.PROP_PHY_ENABLED, 1) # detect baudrate
+
+        # result should not be None for both NCP and RCP
+        result = wpan_api.prop_get_value(SPINEL.PROP_CAPS) # confirm OpenThread Sniffer
+
+        # check whether or not is OpenThread Sniffer
+        stream.close()
+        if result is not None:
+            baudrate = speed
+            break
+
+    if baudrate is not None:
+        if sys.platform == 'win32':
+            # Wireshark only shows the value of key `display`('OpenThread Sniffer').
+            # Here intentionally appends interface in the end (e.g. 'OpenThread Sniffer: COM0').
+            print('interface {value=%s:%s}{display=OpenThread Sniffer %s}' % (interface, baudrate, interface), file=console, flush=True)
+        else:
+            # On Linux or MacOS, wireshark will show the concatenation of the content of `display`
+            # and `interface` by default (e.g. 'OpenThread Sniffer: /dev/ttyACM0').
+            print('interface {value=%s:%s}{display=OpenThread Sniffer}' % (interface, baudrate), file=console, flush=True)
+
+def extcap_interfaces():
+    """List available interfaces to capture from"""
+    console = sys.stdout
+
+    log_file = open(os.path.join(tempfile.gettempdir(), 'extcap_ot_interfaces.log'), 'w')
+    print('extcap {version=1.0.0}{display=OpenThread Sniffer}{help=https://github.com/openthread/pyspinel}')
+
+    for interface in comports():
+        th = threading.Thread(target=serialopen, args=(interface, console, log_file))
+        th.daemon = True
+        th.start()
+    time.sleep(0.5)
```

Comment:
This makes the script to always block for 0.5 seconds. However if all daemon threads quit very quickly, the main thread should not wait for 0.5 seconds, but quit immediately.

Reply:
Any ideas how to do this easily? Does Python have anything like Promise.race in Javascript? Or I have to start another 2 threads, one to wait for 0.5 seconds and the other to wait for all worker threads, then use some fancy condition variables and atomic flags to get notified when any of them returns.

Reply:
No, it does not have to be so complex. Try code like this:
```python
deadline = time.time() + 0.5
for th in threads:
    th.join(timeout=deadline-time.time())

```

Reply:
OK, it works.